packages/Discretization/src/DTK_Interpolation_def.hpp

```diff
@@ -0,0 +1,669 @@
+    // Tpetra::Distributor tmp_2(*(tmp_1.getReverse()));
+    // DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+    //     tmp_2, exported_points, imported_points );
+    //
+    // tmp_2 should be the same as distributed_search_distributor yet the code
+    // hangs. So instead, create a new distributor from scratch. Note that I
+    // cannot reproduce this simple on a simple test case.
+    Kokkos::View<int *, DeviceType> own_rank( "own_rank", ranks.extent( 0 ) );
+    Kokkos::deep_copy( own_rank, _comm->getRank() );
+    Kokkos::View<int *, DeviceType> imported_ranks( "imported_ranks",
+                                                    n_imports );
+    DataTransferKit::DistributedSearchTreeImpl<DeviceType>::sendAcrossNetwork(
+        distributed_search_distributor, own_rank, imported_ranks );
+    auto imported_ranks_host = Kokkos::create_mirror_view( imported_ranks );
+    Kokkos::deep_copy( imported_ranks_host, imported_ranks );
+    _distributor.createFromSends( Teuchos::ArrayView<int const>(
+        imported_ranks_host.data(), imported_ranks_host.extent( 0 ) ) );
+}
+
+template <typename DeviceType>
+void Interpolation<DeviceType>::filterTopology(
+    Kokkos::View<unsigned int *, DeviceType> topo, unsigned int topo_id,
+    Kokkos::View<int *, DeviceType> cell_indices,
+    Kokkos::View<DataTransferKit::Point *, DeviceType> points,
+    Kokkos::View<int *, DeviceType> query_ids,
+    Kokkos::View<unsigned int *, DeviceType> map,
+    Kokkos::View<int *, DeviceType> filtered_cell_indices,
+    Kokkos::View<double **, DeviceType> filtered_points,
+    Kokkos::View<int *, DeviceType> filtered_query_ids )
+{
+    using ExecutionSpace = typename DeviceType::execution_space;
+    unsigned int const size = topo.extent( 0 );
+    Kokkos::View<unsigned int *, DeviceType> offset( "offset", size );
+    computeOffset( topo, topo_id, offset );
+
+    // Create Kokkos::View with the points and the cell indices associated
+    // with cells of topo_id topology. Also transform 3D points back to 2D
+    // points.
+    // We want to use _bounding_box_to_cell in the lambda function but we want
+    // can't because it is a private variable so create a local variable
+    // instead.
+    Kokkos::View<unsigned int **, DeviceType> bounding_box_to_cell =
+        _bounding_box_to_cell;
+    unsigned int dim = _dim;
+    Kokkos::parallel_for(
+        "filter_data", Kokkos::RangePolicy<ExecutionSpace>( 0, size ),
+        KOKKOS_LAMBDA( int const i ) {
+            if ( topo( i ) == topo_id )
+            {
+                unsigned int const k = offset( i );
+                filtered_cell_indices( k ) =
+                    bounding_box_to_cell( cell_indices( i ), topo_id );
+                for ( unsigned int j = 0; j < dim; ++j )
+                {
+                    filtered_points( k, j ) = points( i )[j];
+                    filtered_query_ids( k ) = query_ids( i );
+                    map( k ) = i;
```

Comment:
`map` is also not very descriptive.  Is that used to map the index of the cell in the vector of all cells with mixed topology and map it to the index for the cell in a vector of cells with the same topology?

Reply:
also why are the previous two lines inside the loop over the dimensions?

Reply:
Yes that is what `map` does. The for loop is too big, you are right.